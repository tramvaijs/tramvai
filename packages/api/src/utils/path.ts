import path from 'node:path';

export const resolveAbsolutePathForFile = ({
  file,
  sourceDir,
  rootDir,
}: {
  file: string;
  sourceDir: string;
  rootDir: string;
}): string => {
  if (path.isAbsolute(file)) {
    return file;
  }
  if (path.isAbsolute(sourceDir)) {
    return path.join(sourceDir, file);
  }
  if (path.isAbsolute(rootDir)) {
    return path.join(rootDir, sourceDir, file);
  }
  return path.resolve(rootDir, sourceDir, file);
};

export const resolveAbsolutePathForFolder = ({
  folder,
  rootDir,
}: {
  folder: string;
  rootDir: string;
}): string => {
  if (path.isAbsolute(folder)) {
    return folder;
  }
  if (path.isAbsolute(rootDir)) {
    return path.join(rootDir, folder);
  }
  return path.resolve(rootDir, folder);
};
