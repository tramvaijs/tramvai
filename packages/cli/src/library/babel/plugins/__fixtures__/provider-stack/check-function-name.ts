// @ts-ignore
// eslint-disable-next-line import/no-unresolved
import { Provider } from '@tramvai/core';

export const providers: Provider[] = [
  {
    provide: 'a',
    useFactory: () => {},
  },
];
