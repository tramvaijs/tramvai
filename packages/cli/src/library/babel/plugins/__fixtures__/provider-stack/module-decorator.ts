// @ts-ignore
// eslint-disable-next-line import/no-unresolved
import { module, Scope } from '@tramvai/core';

@module({
  providers: [
    {
      provide: 'provide',
      // eslint-disable-next-line @typescript-eslint/no-extraneous-class
      useClass: class {},
      multi: true,
      deps: {},
    },
    {
      provide: 'test',
      useFactory: () => {},
      multi: false,
      scope: Scope.SINGLETON,
    },
    {
      provide: 'abc',
      useValue: 3,
    },
  ],
})
// eslint-disable-next-line @typescript-eslint/no-extraneous-class
export class ClassModule {}
