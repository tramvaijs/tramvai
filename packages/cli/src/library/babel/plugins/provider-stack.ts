import type { Identifier, Expression, ObjectExpression } from '@babel/types';
import type { NodePath } from '@babel/traverse';
import type { Plugin } from './types.h';

const INNER_COMMENT = '__tramvai_stack_provider__';

export const providerStackPlugin: Plugin = ({ types: t }) => {
  const isProvider = (path: NodePath<ObjectExpression>) => {
    const { properties } = path.node;
    let hasProvideKey = false;
    let hasUseKey = false;

    for (const prop of properties) {
      if (t.isObjectProperty(prop)) {
        const { key } = prop;

        if (
          t.isIdentifier(key, { name: 'useValue' }) ||
          t.isIdentifier(key, { name: 'useFactory' }) ||
          t.isIdentifier(key, { name: 'useClass' })
        ) {
          hasUseKey = true;
        }

        if (t.isIdentifier(key, { name: 'provide' })) {
          hasProvideKey = true;
        }

        if (hasProvideKey && hasUseKey) {
          return true;
        }
      }
    }

    return false;
  };

  const generateStack = (path: NodePath<ObjectExpression>) => {
    if (path.scope.hasBinding('Error', true)) {
      path.scope.rename('Error');
    }

    return t.memberExpression(t.newExpression(t.identifier('Error'), []), t.identifier('stack'));
  };

  const defineProperty = (id: Identifier, value: Expression) => {
    return t.callExpression(
      t.memberExpression(t.identifier('Object'), t.identifier('defineProperty')),
      [
        id,
        t.stringLiteral('__stack'),
        t.objectExpression([
          t.objectProperty(t.identifier('enumerable'), t.booleanLiteral(false)),
          t.objectProperty(t.identifier('value'), value),
        ]),
      ]
    );
  };

  return {
    visitor: {
      ObjectExpression(path) {
        if (
          isProvider(path) &&
          !path.node.leadingComments?.find((comment) => comment.value === INNER_COMMENT)
        ) {
          const ref = path.scope.generateDeclaredUidIdentifier('ref');

          const newNode = t.addComment(
            t.objectExpression(path.node.properties),
            'leading',
            INNER_COMMENT
          );

          path.replaceWith(
            t.sequenceExpression([
              t.assignmentExpression('=', ref, newNode),
              defineProperty(ref, generateStack(path)),
              ref,
            ])
          );
        }
      },
    },
  };
};

export default providerStackPlugin;
