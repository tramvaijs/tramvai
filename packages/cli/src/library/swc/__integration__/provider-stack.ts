import type { Provider } from '@tramvai/core';
import { provide } from '@tramvai/core';
import { TEST_TOKEN } from './create-token-pure';

export const providers: Provider[] = [
  provide({
    provide: TEST_TOKEN,
    useValue: 'test',
  }),
];
