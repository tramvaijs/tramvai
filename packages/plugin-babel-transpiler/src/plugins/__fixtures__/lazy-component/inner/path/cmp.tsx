import React from 'react';

const InnerCmp = () => {
  return <div>inner cmp</div>;
}

export default InnerCmp;
