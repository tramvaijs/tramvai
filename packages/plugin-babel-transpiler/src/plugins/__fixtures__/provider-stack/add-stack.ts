// @ts-ignore
// eslint-disable-next-line import/no-unresolved
import { Provider, Scope } from '@tramvai/core';

export const providers: Provider[] = [
  {
    provide: 'a',
    useValue: 1,
  },
  {
    provide: 'b',
    multi: true,
    deps: {},
    useFactory: () => {},
  },
  {
    provide: 'c',
    scope: Scope.SINGLETON,
    // eslint-disable-next-line @typescript-eslint/no-extraneous-class
    useClass: class {},
  },
];
