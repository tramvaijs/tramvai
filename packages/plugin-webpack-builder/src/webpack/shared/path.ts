import path from 'node:path';

export const resolveAbsolutePathForEntry = ({
  entry,
  sourceDir,
  rootDir,
}: {
  entry: string;
  sourceDir: string;
  rootDir: string;
}): string => {
  if (path.isAbsolute(entry)) {
    return entry;
  }
  if (path.isAbsolute(sourceDir)) {
    return path.join(sourceDir, entry);
  }
  if (path.isAbsolute(rootDir)) {
    return path.join(rootDir, sourceDir, entry);
  }
  return path.resolve(rootDir, sourceDir, entry);
};

export const resolveAbsolutePathForOutput = ({
  output,
  rootDir,
}: {
  output: string;
  rootDir: string;
}): string => {
  if (path.isAbsolute(output)) {
    return output;
  }
  if (path.isAbsolute(rootDir)) {
    return path.join(rootDir, output);
  }
  return path.resolve(rootDir, output);
};
