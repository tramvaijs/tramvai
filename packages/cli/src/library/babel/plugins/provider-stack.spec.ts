import path from 'path';
import pluginTester from 'babel-plugin-tester';
import { providerStackPlugin } from './provider-stack';
import babelConfig from '../index';

pluginTester({
  plugin: providerStackPlugin,
  pluginName: 'provider-stack',
  filename: path.join(__dirname, '__fixtures__', 'provider-stack', 'test'),
  babelOptions: babelConfig({
    typescript: true,
    generateDataQaTag: false,
    loader: false,
  }),
  tests: {
    'do nothing': {
      fixture: 'do-nothing.ts',
      snapshot: true,
    },
    'add __stack property to module decorator providers': {
      fixture: 'module-decorator.ts',
      snapshot: true,
    },
    'add __stack property external providers': {
      fixture: 'add-stack.ts',
      snapshot: true,
    },
    'add __stack property if Error in scope': {
      fixture: 'error-in-scope.ts',
      snapshot: true,
    },
    'inside await': {
      fixture: 'inside-await.ts',
      snapshot: true,
    },
  },
});

pluginTester({
  plugin: providerStackPlugin,
  pluginName: 'provider-stack',
  filename: path.join(__dirname, '__fixtures__', 'provider-stack', 'test'),
  babelOptions: babelConfig({
    typescript: true,
    generateDataQaTag: false,
    loader: false,
    excludesPresetEnv: ['transform-function-name'],
  }),
  tests: {
    'check function naming': {
      fixture: 'check-function-name.ts',
      snapshot: true,
    },
  },
});
