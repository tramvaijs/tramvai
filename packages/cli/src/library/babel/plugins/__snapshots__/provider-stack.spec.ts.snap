// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`provider-stack add __stack property external providers: add __stack property external providers 1`] = `

// @ts-ignore
// eslint-disable-next-line import/no-unresolved
import { Provider, Scope } from '@tramvai/core';

export const providers: Provider[] = [
  {
    provide: 'a',
    useValue: 1,
  },
  {
    provide: 'b',
    multi: true,
    deps: {},
    useFactory: () => {},
  },
  {
    provide: 'c',
    scope: Scope.SINGLETON,
    // eslint-disable-next-line @typescript-eslint/no-extraneous-class
    useClass: class {},
  },
];

      ↓ ↓ ↓ ↓ ↓ ↓

var _ref, _ref2, _ref3;
// @ts-ignore
// eslint-disable-next-line import/no-unresolved
import { Scope } from '@tramvai/core';
export var providers = [
  ((_ref =
    /*__tramvai_stack_provider__*/
    {
      provide: 'a',
      useValue: 1,
    }),
  Object.defineProperty(_ref, '__stack', {
    enumerable: false,
    value: new Error().stack,
  }),
  _ref),
  ((_ref2 =
    /*__tramvai_stack_provider__*/
    {
      provide: 'b',
      multi: true,
      deps: {},
      useFactory: function useFactory() {},
    }),
  Object.defineProperty(_ref2, '__stack', {
    enumerable: false,
    value: new Error().stack,
  }),
  _ref2),
  ((_ref3 =
    /*__tramvai_stack_provider__*/
    {
      provide: 'c',
      scope: Scope.SINGLETON,
      // eslint-disable-next-line @typescript-eslint/no-extraneous-class
      useClass: function useClass() {},
    }),
  Object.defineProperty(_ref3, '__stack', {
    enumerable: false,
    value: new Error().stack,
  }),
  _ref3),
];


`;

exports[`provider-stack add __stack property if Error in scope: add __stack property if Error in scope 1`] = `

// @ts-ignore
// eslint-disable-next-line import/no-unresolved
import { Provider, Scope, Error } from '@tramvai/core';

export const providers: Provider[] = [
  {
    provide: 'a',
    useValue: 1,
  },
  {
    provide: 'b',
    multi: true,
    deps: {},
    useFactory: () => {
      console.log(Error);
    },
  },
  {
    provide: 'c',
    scope: Scope.SINGLETON,
    // eslint-disable-next-line @typescript-eslint/no-extraneous-class
    useClass: class {},
  },
];

      ↓ ↓ ↓ ↓ ↓ ↓

var _ref, _ref2, _ref3;
// @ts-ignore
// eslint-disable-next-line import/no-unresolved
import { Scope, Error as _Error } from '@tramvai/core';
export var providers = [
  ((_ref =
    /*__tramvai_stack_provider__*/
    {
      provide: 'a',
      useValue: 1,
    }),
  Object.defineProperty(_ref, '__stack', {
    enumerable: false,
    value: new Error().stack,
  }),
  _ref),
  ((_ref2 =
    /*__tramvai_stack_provider__*/
    {
      provide: 'b',
      multi: true,
      deps: {},
      useFactory: function useFactory() {
        console.log(_Error);
      },
    }),
  Object.defineProperty(_ref2, '__stack', {
    enumerable: false,
    value: new Error().stack,
  }),
  _ref2),
  ((_ref3 =
    /*__tramvai_stack_provider__*/
    {
      provide: 'c',
      scope: Scope.SINGLETON,
      // eslint-disable-next-line @typescript-eslint/no-extraneous-class
      useClass: function useClass() {},
    }),
  Object.defineProperty(_ref3, '__stack', {
    enumerable: false,
    value: new Error().stack,
  }),
  _ref3),
];


`;

exports[`provider-stack add __stack property to module decorator providers: add __stack property to module decorator providers 1`] = `

// @ts-ignore
// eslint-disable-next-line import/no-unresolved
import { module, Scope } from '@tramvai/core';

@module({
  providers: [
    {
      provide: 'provide',
      // eslint-disable-next-line @typescript-eslint/no-extraneous-class
      useClass: class {},
      multi: true,
      deps: {},
    },
    {
      provide: 'test',
      useFactory: () => {},
      multi: false,
      scope: Scope.SINGLETON,
    },
    {
      provide: 'abc',
      useValue: 3,
    },
  ],
})
// eslint-disable-next-line @typescript-eslint/no-extraneous-class
export class ClassModule {}

      ↓ ↓ ↓ ↓ ↓ ↓

var _dec, _ref, _ref2, _ref3, _class;
// @ts-ignore
// eslint-disable-next-line import/no-unresolved
import { module, Scope } from '@tramvai/core';
export var ClassModule =
  ((_dec = module({
    providers: [
      ((_ref =
        /*__tramvai_stack_provider__*/
        {
          provide: 'provide',
          // eslint-disable-next-line @typescript-eslint/no-extraneous-class
          useClass: function useClass() {},
          multi: true,
          deps: {},
        }),
      Object.defineProperty(_ref, '__stack', {
        enumerable: false,
        value: new Error().stack,
      }),
      _ref),
      ((_ref2 =
        /*__tramvai_stack_provider__*/
        {
          provide: 'test',
          useFactory: function useFactory() {},
          multi: false,
          scope: Scope.SINGLETON,
        }),
      Object.defineProperty(_ref2, '__stack', {
        enumerable: false,
        value: new Error().stack,
      }),
      _ref2),
      ((_ref3 =
        /*__tramvai_stack_provider__*/
        {
          provide: 'abc',
          useValue: 3,
        }),
      Object.defineProperty(_ref3, '__stack', {
        enumerable: false,
        value: new Error().stack,
      }),
      _ref3),
    ],
  })),
  _dec((_class = function ClassModule() {})) || _class);


`;

exports[`provider-stack check function naming: check function naming 1`] = `

// @ts-ignore
// eslint-disable-next-line import/no-unresolved
import { Provider } from '@tramvai/core';

export const providers: Provider[] = [
  {
    provide: 'a',
    useFactory: () => {},
  },
];

      ↓ ↓ ↓ ↓ ↓ ↓

var _ref;
// @ts-ignore
// eslint-disable-next-line import/no-unresolved

export var providers = [
  ((_ref =
    /*__tramvai_stack_provider__*/
    {
      provide: 'a',
      useFactory: function () {},
    }),
  Object.defineProperty(_ref, '__stack', {
    enumerable: false,
    value: new Error().stack,
  }),
  _ref),
];


`;

exports[`provider-stack do nothing: do nothing 1`] = `

export const providers = [
  {
    provide: 'test',
  },
  {
    provide: 'a',
    a: 'a',
  },
  {
    useClass: 'test',
    b: 'b',
  },
  {
    useValue: 'c',
  },
];

      ↓ ↓ ↓ ↓ ↓ ↓

export var providers = [
  {
    provide: 'test',
  },
  {
    provide: 'a',
    a: 'a',
  },
  {
    useClass: 'test',
    b: 'b',
  },
  {
    useValue: 'c',
  },
];


`;

exports[`provider-stack inside await: inside await 1`] = `

export const func = async () => {
  await Promise.resolve([
    {
      provide: 'a',
      useValue: 'a',
    },
    {
      provide: 'b',
      useValue: 'b',
    },
  ]);
}

      ↓ ↓ ↓ ↓ ↓ ↓

import _asyncToGenerator from '@babel/runtime/helpers/esm/asyncToGenerator';
import _regeneratorRuntime from '@babel/runtime/regenerator';
export var func = /*#__PURE__*/ (function () {
  var _ref = _asyncToGenerator(
    /*#__PURE__*/ _regeneratorRuntime.mark(function _callee() {
      var _ref2, _ref3;
      return _regeneratorRuntime.wrap(function _callee$(_context) {
        while (1)
          switch ((_context.prev = _context.next)) {
            case 0:
              _context.next = 2;
              return Promise.resolve([
                ((_ref2 =
                  /*__tramvai_stack_provider__*/
                  {
                    provide: 'a',
                    useValue: 'a',
                  }),
                Object.defineProperty(_ref2, '__stack', {
                  enumerable: false,
                  value: new Error().stack,
                }),
                _ref2),
                ((_ref3 =
                  /*__tramvai_stack_provider__*/
                  {
                    provide: 'b',
                    useValue: 'b',
                  }),
                Object.defineProperty(_ref3, '__stack', {
                  enumerable: false,
                  value: new Error().stack,
                }),
                _ref3),
              ]);
            case 2:
            case 'end':
              return _context.stop();
          }
      }, _callee);
    })
  );
  return function func() {
    return _ref.apply(this, arguments);
  };
})();


`;
